Add service manager

brb opened the pull request:
This PR introduces the service manager which handles all the plumbing of services between cilium-agent and the BPF maps.

Previously, the manager functionality was spread among `pkg/loadbalancer/loadbalancer.go`, `daemon/loadbalancer.go` and `pkg/lbmap/bpfservice.go`. Now, it resides in a single module - `pkg/service/service.go`.

I didn't add any test (yet), as this PR is big enough, and I didn't want to introduce more changes by adding the `lbmap` interface. Going to do in a subsequent PR. For now, I'm relying on the integration tests. 

Next steps:

- Unit tests for the service manager.
- `TODO(brb)` items.
- Cleanup types mess in `pkg/loadbalancer`.
- Get rid of some methods in `pkg/lbmap`.

Reviewable per commit.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/9274)
<!-- Reviewable:end -->

brb committed:
lbmap: Add RevNatValue method to ServiceKeyV2

It's going to be used when updating service, and it will allow
us to get rid of the legacy Service{Key,Value} types soon.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
counter: Add reference counting for strings

This is a copy-paste from pkg/counter/interger.go.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
lbmap: Add stateless functions for LB BPF maps manipulation

This commit introduces functions to add, delete, update entries
in the lb{4,6}_backend, lb{4,6}_service_v2 and lb{4,6}_reverse_nat
BPF maps.

The functions do not update and do not rely on the lbmap cache
(bpfservice.go).

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
Add temporary package for service management

This commit introduces the pkg/svc package.

The package is responsible for managing services which involves:

- Triggering Allocation / release of service and backend IDs.
- Storing services in local cache.
- Reflecting changes into the BPF LB maps.

This package merges functionality which previously was spread
among pkg/lbmap/bpfservice.go, daemon/loadbalancer.go,
pkg/loadbalancer/loadbalancer.go modules.

The package is temporary and all its code will be moved to
pkg/service after the potential circular dependencies have
been eliminated.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
daemon: Remove SyncLBMap invocation

Previously, SyncLBMap() was used to restore the daemon.loadBalancer
cache from the LB BPF maps, and to remove a SVC if its ID has changed
when running in non-k8s env and with KVStore enabled for SVC ID
allocation.

As the cache will be restored by pkg/svc.RestoreServices() and
the KVStore is no longer used for the ID allocation, we can remove
the call to SyncLBMap().

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
daemon: Use pkg/svc for service management

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
Remove dead code

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
bpf: Get rid of dependency on pks/k8s

This helps to eliminate one of the circular dependencies when
pkg/svc is moved into pkg/service.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
service: Move pkg/svc/svc.go to pkg/service/service.go

Also, eliminate the circular dependency by passing the callback
to service.SyncWithK8s() instead of k8s service frontend values.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

aanm commented on pkg/svc/svc.go:
```diff
@@ -0,0 +1,486 @@
+		}
+	}
+
+	for _, m := range toOpen {
```
Shouldn't we delete first? If `restore` is false won't we "OpenOrCreate" a map and then delete it?

brb replied:
No, we cannot delete entries if the maps haven't been opened.

Previously, we did "OpenOrCreate" (https://github.com/cilium/cilium/blob/master/daemon/datapath.go#L575), and only afterwards "DeleteAll" (https://github.com/cilium/cilium/blob/master/daemon/datapath.go#L601).

aanm replied:
ah it's `DeleteAll` I read it as `DeleteMap`

aanm commented on pkg/svc/svc.go:
```diff
@@ -0,0 +1,486 @@
+	return nil
+}
+
+func (s *Service) updateBackendsCacheLocked(svc *lb.LBSVC, backends []lb.LBBackEnd) ([]lb.LBBackEnd, []lb.BackendID, error) {
```
Some unit tests for these functions would be :100:

brb replied:
It's coming today (albeit in a separate PR)!

ianvernon commented on daemon/loadbalancer.go:
```diff
@@ -184,7 +98,7 @@ func NewDeleteServiceIDHandler(d *Daemon) DeleteServiceIDHandler {
 	log.WithField(logfields.Params, logfields.Repr(params)).Debug("DELETE /service/{id} request")
 
-	found, err := h.d.svcDeleteByID(loadbalancer.ServiceID(params.ID))
+	found, err := h.d.svc.DeleteServiceByID(loadbalancer.ServiceID(params.ID))
```
Instead of having `deleteServiceID` have a `*Daemon`, it should just have a `*svc.Service` member. That way we can just do `h.svc.DeleteServiceByID` - we could use this to make it easier to test our API handler functions in the future, for instance.

brb replied:
Neat idea! Going to change in all `daemon/loadbalancer.go` handlers.

ianvernon commented on daemon/loadbalancer.go:
```diff
@@ -256,10 +124,7 @@ func (h *getServiceID) Handle(params GetServiceIDParams) middleware.Responder {
-	defer d.loadBalancer.BPFMapMU.RUnlock()
-
-	if svc, ok := d.loadBalancer.SVCMapID[loadbalancer.ServiceID(params.ID)]; ok {
+	if svc, ok := d.svc.GetDeepCopyServiceByID(loadbalancer.ServiceID(params.ID)); ok {
```
Same here.

ianvernon commented on daemon/loadbalancer.go:
```diff
@@ -279,277 +144,12 @@ func (h *getService) Handle(params GetServiceParams) middleware.Responder {
-		"failed":           failed,
-		"skipped":          skipped,
-		"removed":          removed,
-	}).Info("Restore service IDs from BPF maps")
```
Seeing this much code deleted from `daemon/` is *awesome*!

brb committed:
daemon: Inline SVCAdd method into PUT handler

This is required, as we are going to get rid of the *Daemon instance
dependency in the LB handler.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
daemon: Get rid of Daemon dependency in LB API handler

This will ease the testing of the handler.

Signed-off-by: Martynas Pumputis <m@lambda.lt>
